### ISSUE TITLE
AWS Signature V4

### ISSUE THREAD

**Issue by kevin-yen** [JuliaCloud/AWSS3.jl#89]
Hi, I encountered an error when using the signed URL generated by the `s3_sign_url` function. The response to the signed URL is this

```xml
<Error>
  <Code>InvalidArgument</Code>
  <Message>Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.</Message>
  <ArgumentName>Authorization</ArgumentName>
  <ArgumentValue>null</ArgumentValue>
  <RequestId> ... the request ID ... </RequestId>
  <HostId>... the host ID ...</HostId>
</Error>
```

Correct me if I'm wrong, but from the code it looks like we're using AWS signature V2? https://github.com/JuliaCloud/AWSS3.jl/blob/c881edf4318dd0e78de93b0b78f3baa440e0e0cb/src/AWSS3.jl#L777-L800
Can we update the algorithm to use V4?
https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html

Maybe we can have a keyword parameter like `signature_version` that defaults to v4?

**Comment by mattBrzezinski** [comment]
Hi Kevin,

We should probably move to `v4` signing. Which AWS region are you using when encountering this error? The `v2` signing should still be working in:

- us-east-1 (N. Virginia)
- us-west-1 (N. California) / us-west-2 (Oregon)
- eu-west-1 (Ireland)
- ap-northeast-1 (Tokyo)
- ap-southeast-1 (Singapore)
- ap-southeast-2 (Sydney)
- sa-east-1 (Sāo Paulo)

I like the idea of `signature_version` as a kwarg.

**Comment by kevin-yen** [comment]
I'm using us-west-2.

The weird part is on our qa bucket it works fine but prod bucket it doesn't work. I'm guessing its configuration differences in how the cloud team at my company set up the buckets.

I'm working to implement V4 because I need it for work. I can send a PR once it's done

**Comment by mattBrzezinski** [comment]
Sounds good, when I see it come through I can prioritize reviewing it to unblock you!

**Comment by kevin-yen** [comment]
@mattBrzezinski I created the PR https://github.com/JuliaCloud/AWSS3.jl/pull/90. I tested it by writing a separate function in my application code and made sure it worked in our qa and prod environments. Then I copied the function as `_s3_sign_url_v4`. I would like to run the unit tests in this repo but I wasn't able to run them because of this error

```
AWSS3.jl: Error During Test at C:\Users\Kevin Yen\git\github\AWSS3.jl\test\runtests.jl:14
  Got exception outside of a @test
  LoadError: IOError(Base.IOError("connect: network is unreachable (ENETUNREACH)", -4062) during request(http://169.254.169.254/latest/meta-data/iam/info))
```

Can you help me with unit testing my changes?

Let me know if there are additional changes you want to see to the PR

**Comment by kevin-yen** [comment]
@mattBrzezinski 
In the course of coding the v4 signature process. I also encountered

```xml
<Error>
  <Code>AccessDenied</Code>
  <Message>Requests specifying Server Side Encryption with AWS KMS managed keys must be made over a secure connection.</Message>
  <ArgumentName>x-amz-server-side-encryption</ArgumentName>
  <ArgumentValue>null</ArgumentValue>
  <RequestId>... request ID ...</RequestId>
  <HostId>... host ID ... </HostId>
</Error>
```

I was able to resolve it by setting `protocol = "https"`.

However, I feel we should be defaulting to the more secure connection. It should be a deliberate choice to set it to insecure "http" so I'm curious about your opinion to change the default to "https"

**Comment by mattBrzezinski** [comment]
> @mattBrzezinski I created the PR #90. I tested it by writing a separate function v4 signing code in my application code and made sure it worked in our qa and prod environments. Then I copied the function as `_s3_sign_url_v4` into this library.
> 
> I would like to test the code with the existing code you have in this repo but I wasn't able to run them because of this error
> 
> ```
> AWSS3.jl: Error During Test at C:\Users\Kevin Yen\git\github\AWSS3.jl\test\runtests.jl:14
>   Got exception outside of a @test
>   LoadError: IOError(Base.IOError("connect: network is unreachable (ENETUNREACH)", -4062) during request(http://169.254.169.254/latest/meta-data/iam/info))
> ```
> 
> Can you help me with unit testing my changes? I want to make sure the code is up to your standards and tested.
> 
> Also let me know if there are additional changes you want to see to the PR

Just saw this after I had submitted a review. The `Base.IOError` is because that URL is only accessible by EC2 instances. I can take a look at what is happening tomorrow and post an update!

**Comment by mattBrzezinski** [comment]
I just realized that this code for signing v4 signatures has been written, [here](https://github.com/JuliaCloud/AWSCore.jl/blob/master/src/sign.jl#L64)

**Comment by kevin-yen** [comment]
Oh, that's a good find! Let's integrate with that then. I'm not sure whether I have time tomorrow to look into this further but I might have some time later in the week to use the preexisting one

**Comment by kevin-yen** [comment]
It looks like we'll need to modify or add to the existing code so it generates a pre-signed URL instead of an authenticated request

**Comment by kevin-yen** [comment]
@mattBrzezinski Since we didn't end up using the v4 signing code from AWSCore. I've created another issue for that https://github.com/JuliaCloud/AWSS3.jl/issues/92 and closing this one. Thanks for merging my PR
